src/server/server_family.cc

```diff
@@ -114,12 +118,73 @@ void SubstituteFilenameTsPlaceholder(fs::path* filename, std::string_view replac
   *filename = absl::StrReplaceAll(filename->string(), {{"{timestamp}", replacement}});
 }
 
-string InferLoadFile(fs::path data_dir) {
+bool IsCloudPath(string_view path) {
+  return absl::StartsWith(path, "s3://");
+}
+
+// Returns bucket_name, obj_path for an s3 path.
+optional<pair<string_view, string_view>> GetBucketPath(string_view path) {
+  string_view clean = absl::StripPrefix(path, "s3://");
+
+  size_t pos = clean.find('/');
+  if (pos == string_view::npos)
+    return nullopt;
+
+  string_view bucket_name = clean.substr(0, pos);
+  string_view obj_path = clean.substr(pos + 1);
+  return make_pair(bucket_name, obj_path);
+}
+
+string InferLoadFile(string_view dir, cloud::AWS* aws) {
+  fs::path data_folder;
+  string_view bucket_name, obj_path;
+
+  if (dir.empty()) {
+    data_folder = fs::current_path();
+  } else {
+    if (IsCloudPath(dir)) {
+      CHECK(aws);
+      auto res = GetBucketPath(dir);
+      if (!res) {
+        LOG(ERROR) << "Invalid S3 path: " << dir;
+        return {};
+      }
+      data_folder = dir;
+      bucket_name = res->first;
+      obj_path = res->second;
+    } else {
+      error_code file_ec;
+      data_folder = fs::canonical(dir, file_ec);
+      if (file_ec) {
+        LOG(ERROR) << "Data directory error: " << file_ec.message();
+        return {};
+      }
+    }
+  }
+
+  LOG(INFO) << "Data directory is " << data_folder;
+
   const auto& dbname = GetFlag(FLAGS_dbfilename);
   if (dbname.empty())
     return string{};
 
-  fs::path fl_path = data_dir.append(dbname);
+  if (IsCloudPath(dir)) {
+    cloud::S3Bucket bucket(*aws, bucket_name);
+    ProactorBase* proactor = shard_set->pool()->GetNextProactor();
+    auto ec = proactor->Await([&] { return bucket.Connect(kBucketConnectMs); });
+    if (ec) {
+      LOG(ERROR) << "Couldn't connect to S3 bucket: " << ec.message();
+      return {};
+    }
+
+    fs::path fl_path{obj_path};
+    fl_path /= dbname;
```

Comment:
Nit: I'm not a big fan of using `operator /=` for *appending* paths (I'd use `.append()` instead), but that's very much a matter of preference